tests/unit/s2n_server_psk_extension_test.c

```diff
@@ -0,0 +1,416 @@
+
+            uint16_t chosen_psk_wire_index = 0;
+            EXPECT_SUCCESS(s2n_stuffer_read_uint16(&out, &chosen_psk_wire_index));
+            EXPECT_EQUAL(chosen_psk_wire_index, server_conn->psk_params.chosen_psk_wire_index);
+
+            EXPECT_SUCCESS(s2n_connection_free(server_conn));
+            EXPECT_SUCCESS(s2n_connection_free(client_conn));
+            EXPECT_SUCCESS(s2n_stuffer_free(&out));
+        }
+    }
+
+    /* Test: s2n_server_psk_recv */
+    {
+        s2n_extension_type_id key_share_ext_id;
+        EXPECT_SUCCESS(s2n_extension_supported_iana_value_to_id(TLS_EXTENSION_KEY_SHARE, &key_share_ext_id));
+
+        /* Test s2n_server_psk_recv for invalid TLS versions <= TLS1.2 */
+        {
+            struct s2n_stuffer out = { 0 };
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&out, 0));
+
+            struct s2n_connection *conn = NULL;
+            EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+
+            EXPECT_OK(setup_psks(conn));
+
+            /* The keyshare extension needs to be present, as s2n currently only
+             * supports pre-shared keys in (EC)DHE key exchange mode.
+             */
+            S2N_CBIT_SET(conn->extension_requests_received, key_share_ext_id);
+
+            uint16_t chosen_psk_wire_index = 1;
+            /* Incorrect protocol version */
+            conn->actual_protocol_version = S2N_TLS12;
+            EXPECT_SUCCESS(s2n_stuffer_write_uint16(&out, chosen_psk_wire_index));
+
+            EXPECT_NULL(conn->psk_params.chosen_psk);
+            EXPECT_SUCCESS(s2n_server_psk_extension.recv(conn, &out));
+            EXPECT_NULL(conn->psk_params.chosen_psk);
+
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+            EXPECT_SUCCESS(s2n_stuffer_free(&out));
+        }
+
+        /* Test s2n_server_psk_recv when server key_share extension is not present */
+        {
+            struct s2n_stuffer out = { 0 };
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&out, 0));
+
+            struct s2n_connection *conn = NULL;
+            EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+
+            EXPECT_OK(setup_psks(conn));
+
+            uint16_t chosen_psk_wire_index = s2n_array_len(test_psks) - 1;
+            conn->actual_protocol_version = S2N_TLS13;
+
+            EXPECT_SUCCESS(s2n_stuffer_write_uint16(&out, chosen_psk_wire_index));
+
+            EXPECT_NULL(conn->psk_params.chosen_psk);
+            EXPECT_FAILURE_WITH_ERRNO(s2n_server_psk_extension.recv(conn, &out), S2N_ERR_MISSING_EXTENSION);
+            EXPECT_NULL(conn->psk_params.chosen_psk);
+
+            /* Validate that the client list of PSKs is not wiped */
+            for (size_t i = 0; i < conn->psk_params.psk_list.len; i++) {
+                struct s2n_psk *psk = NULL;
+                EXPECT_OK(s2n_array_get(&conn->psk_params.psk_list, i, (void**)&psk)); 
+                EXPECT_OK(validate_psk_is_not_wiped(psk));
+            }
```

Comment:
Do we define this behavior on an error, or does it just happen to be this way? We shouldn't enforce undefined behavior. Same for your other tests where recv errors.

Reply:
Yes, we expect the PSKs to be not wiped if we error here, the PSKs  not chosen are wiped only when the chosen PSK is set. This logic will be cleaned up once we move the wiping on unused PSKs into the key schedule as we discussed offline.